### PR TITLE
New alerts

### PR DESCRIPTION
**Pull request by Eabramov84**

### CODE REVIEW

File: alerts/alerts_test.go [hunk 1]
```diff
@@ -332,122 +321,62 @@ func TestMessageFilter(t *testing.T) {
...
-		alert.Timestamp = validationAlert.Timestamp
-		if !proto.Equal(alert, &validationAlert) {
-			return false, aoserrors.New("received validation alert != send alert")
-		}
-
-		return true, nil
-	}); err != nil {
-		t.Errorf("Result failed: %s", err)
+	if len(alertsHandler.GetAlertsChannel()) != 50 {
+		t.Error("incorrect channel size ", len(alertsHandler.GetAlertsChannel()))
```
Comment (al1img) [inline alerts/alerts_test.go:363]: `Incorrect...`

File: alerts/alerts_test.go [hunk 2]
```diff
@@ -332,122 +321,62 @@ func TestMessageFilter(t *testing.T) {
...
 	}
 }
```
Comment (al1img) [inline alerts/alerts_test.go:366]: Would be nice to check if API in `setupJournal` is set correctly. For example, PRIORITY is set according to configured version, etc.

File: alerts/alerts.go [hunk 3]
```diff
@@ -60,14 +52,13 @@ const alertChannelSize = 50
 
 const microSecondsinSecond = 1000000
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts.go:53]: fixed

File: alerts/alerts.go [hunk 4]
```diff
@@ -60,14 +52,13 @@ const alertChannelSize = 50
...
- ******************************************************************************/
+ **********************************************************************************************************************/
 
-// ServiceProvider provides service info.
-type ServiceProvider interface {
-	GetService(serviceID string) (service launcher.Service, err error)
-	GetServiceByUnitName(unitName string) (service launcher.Service, err error)
+// InstanceProvider provides instance info.
+type InstanceProvider interface {
+	GetInstanceInfoByID(string) (instance cloudprotocol.InstanceIdent, aosVersion uint64, err error)
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts.go:61]: done

File: alerts/alerts.go [hunk 5]
```diff
@@ -76,42 +67,58 @@ type CursorStorage interface {
...
-	cursorStorage   CursorStorage
-	serviceProvider ServiceProvider
-	filterRegexp    []*regexp.Regexp
+	alertsChannel    chan cloudprotocol.AlertItem
+	config           config.Alerts
+	cursorStorage    CursorStorage
+	instanceProvider InstanceProvider
+	filterRegexp     []*regexp.Regexp
 
 	sync.Mutex
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts.go:91]: done

File: alerts/alerts.go [hunk 6]
```diff
@@ -76,42 +67,58 @@ type CursorStorage interface {
...
-	"aos-communicationmanager.service",
+var coreComponents = []string{ // nolint:gochecknoglobals
+	"aos-servicemanager",
+	"aos-updatemanager",
+	"aos-iamanager",
+	"aos-communicationmanager",
 }
 
-/*******************************************************************************
+var CurrentJournal JournalInterface // nolint:gochecknoglobals
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts.go:112]: done

File: alerts/alerts_test.go [hunk 7]
```diff
@@ -58,54 +50,46 @@ func init() {
...
 
-/*******************************************************************************
- * Vars
- ******************************************************************************/
-
-var systemd *dbus.Conn
+type testSystemdJournal struct {
+	Messages       []*sdjournal.JournalEntry
+	CurrentMessage int
+	sync.RWMutex
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts_test.go:68]: done

File: alerts/alerts_test.go [hunk 8]
```diff
@@ -275,39 +261,42 @@ func TestMessageFilter(t *testing.T) {
...
 					return false, aoserrors.New("wrong alert type")
 				}
 
+				systemAlert, ok := alert.Payload.(cloudprotocol.SystemAlert)
+				if !ok {
+					return false, aoserrors.New("wrong alert type content")
+				}
+
 				if systemAlert.Message != validMessage {
 					return false, aoserrors.New("Receive unexpected alert mesage")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts_test.go:301]: done

File: alerts/alerts_test.go (hunk 1 above)
Comment (Eabramov84) [reply alerts/alerts_test.go:363]: fixed

File: alerts/alerts_test.go [hunk 9]
```diff
@@ -114,158 +98,160 @@ func TestGetSystemError(t *testing.T) {
...
-	if err = sysLog.Notice(messages[len(messages)-1]); err != nil {
-		t.Errorf("Can't write to syslog: %s", err)
-	}
+		journalEntry.Fields[sdjournal.SD_JOURNAL_FIELD_SYSTEMD_UNIT] = "someSystemService"
+		journalEntry.Fields[sdjournal.SD_JOURNAL_FIELD_MESSAGE] = messages[i]
 
-	messages = append(messages, uuid.New().String())
-	if err = sysLog.Info(messages[len(messages)-1]); err != nil {
-		t.Errorf("Can't write to syslog: %s", err)
+		testJournal.Messages = append(testJournal.Messages, &journalEntry)
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts_test.go:121]: Added, but public Method AddMessage

File: alerts/alerts.go [hunk 10]
```diff
@@ -448,3 +350,68 @@ func (instance *Alerts) pushAlert(alert *pb.Alert) {
 
 	instance.alertsChannel <- alert
 }
+
+func (instance *Alerts) fillServiceInstanceAlert(
+	alert *cloudprotocol.AlertItem, entry *sdjournal.JournalEntry, unitName string,
+) {
+	if strings.Contains(unitName, aosServicePrefix) {
+		instanceID := filepath.Base(unitName)
+		instanceID = strings.ReplaceAll(instanceID, aosServicePrefix, "")
```
Comment (al1img) [inline alerts/alerts.go:359]: strings.TrimPrefix

File: alerts/alerts.go [hunk 11]
```diff
@@ -448,3 +350,68 @@ func (instance *Alerts) pushAlert(alert *pb.Alert) {
...
+		instanceID = strings.ReplaceAll(instanceID, ".service", "")
```
Comment (al1img) [inline alerts/alerts.go:360]: strings.TrimSuffix

File: alerts/alerts_test.go [hunk 12]
```diff
@@ -114,158 +99,132 @@ func TestGetSystemError(t *testing.T) {
...
-
-		time.Sleep(500 * time.Millisecond)
+		testJournal.addMessage(messages[i], "someSystemService", "", "3")
 	}
 
-	if err = waitAlerts(alertsHandler.GetAlertsChannel(), 5*time.Second,
-		alerts.AlertTagSystemError, "system", 0, messages); err != nil {
-		t.Errorf("Result failed: %s", err)
-	}
+	time.Sleep(500 * time.Millisecond)
```
Comment (al1img) [inline alerts/alerts_test.go:120]: Do you need this sleep?

File: alerts/alerts_test.go [hunk 13]
```diff
@@ -58,54 +50,46 @@ func init() {
...
 	cursor string
 }
 
-/*******************************************************************************
- * Vars
- ******************************************************************************/
-
-var systemd *dbus.Conn
+type testSystemdJournal struct {
+	Messages       []*sdjournal.JournalEntry
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts_test.go]: moved to private

File: alerts/alerts_test.go (hunk 2 above)
Comment (Eabramov84) [reply alerts/alerts_test.go:375]: done

File: alerts/alerts_test.go (hunk 12 above)
Comment (Eabramov84) [reply alerts/alerts_test.go:120]: removed

File: alerts/alerts_test.go [hunk 14]
```diff
@@ -460,44 +398,77 @@ func (cursorStorage *testCursorStorage) GetJournalCursor() (cursor string, err e
...
+	if len(journal.messages) == 0 {
+		return entry, nil
+	}
+
+	if journal.currentMessage >= len(journal.messages) {
+		return entry, nil
 	}
 
+	entry = journal.messages[journal.currentMessage]
+	journal.currentMessage++
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts_test.go:425]: done

File: alerts/alerts.go (hunk 10 above)
Comment (Eabramov84) [reply alerts/alerts.go:359]: changed

File: alerts/alerts.go (hunk 11 above)
Comment (Eabramov84) [reply alerts/alerts.go:360]: done

File: alerts/alerts_test.go [hunk 15]
```diff
@@ -332,122 +271,103 @@ func TestMessageFilter(t *testing.T) {
...
-		},
-	}
+	for _, etalonMatch := range etalonMatches {
+		exist := false
 
-	if err = waitResult(alertsHandler.GetAlertsChannel(), 5, func(alert *pb.Alert) (success bool, err error) {
-		alert.Timestamp = validationAlert.Timestamp
-		if !proto.Equal(alert, &validationAlert) {
-			return false, aoserrors.New("received validation alert != send alert")
+		for _, journalMatch := range testJournal.systemdMatches {
```
Comment (al1img) [inline alerts/alerts_test.go:345]: Shorter version:

```go
    matchLoop: 
	for _, etalonMatch := range etalonMatches {
		for _, journalMatch := range testJournal.systemdMatches {
			if etalonMatch == journalMatch {
                             continue mathcLoop
			}
		}

	        t.Errorf("Journal filter doesn't contains: %s", etalonMatch)
	}
```

File: alerts/alerts_test.go [hunk 16]
```diff
@@ -460,44 +380,91 @@ func (cursorStorage *testCursorStorage) GetJournalCursor() (cursor string, err e
 	return cursorStorage.cursor, nil
 }
 
-/*******************************************************************************
- * Private
- ******************************************************************************/
+func (journal *testSystemdJournal) Next() (uint64, error) {
+	journal.RLock()
```
Comment (al1img) [inline alerts/alerts_test.go:384]: Should be Lock as you modify message counter.

File: alerts/alerts_test.go [hunk 17]
```diff
@@ -460,44 +380,91 @@ func (cursorStorage *testCursorStorage) GetJournalCursor() (cursor string, err e
...
+	return uint64(sdjournal.SD_JOURNAL_APPEND), nil
 }
 
-func cleanup() {
-	for _, service := range serviceProvider.services {
-		if err := stopService(service.ID); err != nil {
-			log.Errorf("Can't stop service: %s", err)
-		}
+func (journal *testSystemdJournal) GetEntry() (entry *sdjournal.JournalEntry, err error) {
+	journal.Lock()
```
Comment (al1img) [inline alerts/alerts_test.go:401]: Should be RLock as you only read here.

File: alerts/alerts_test.go [hunk 18]
```diff
@@ -460,44 +380,91 @@ func (cursorStorage *testCursorStorage) GetJournalCursor() (cursor string, err e
...
+func (journal *testSystemdJournal) AddMatch(match string) error {
+	journal.systemdMatches = append(journal.systemdMatches, match)
+
+	return nil
+}
+
+func (journal *testSystemdJournal) AddDisjunction() error { return nil }
+
+func (journal *testSystemdJournal) SeekTail() error { return nil }
+func (journal *testSystemdJournal) Previous() (uint64, error) {
```
Comment (al1img) [inline alerts/alerts_test.go:437]: Empty line between function.

File: alerts/alerts_test.go [hunk 19]
```diff
@@ -114,158 +99,118 @@ func TestGetSystemError(t *testing.T) {
...
-		t.Errorf("Result failed: %s", err)
-	}
-}
+	// msg 1
+	testJournal.addMessage("starting", "init.scope", "", "4")
 
-func TestGetServiceManagerAlerts(t *testing.T) {
-	const numMessages = 5
+	// msg 2
+	message := unitName + ": Main process exited, code=dumped, status=11/SEGV"
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts_test.go:159]: There are completely different parameters for the journal message

File: alerts/alerts_test.go (hunk 15 above)
Comment (Eabramov84) [reply alerts/alerts_test.go:345]: thanks

File: alerts/alerts_test.go (hunk 17 above)
Comment (Eabramov84) [reply alerts/alerts_test.go:401]: done

File: alerts/alerts_test.go (hunk 18 above)
Comment (Eabramov84) [reply alerts/alerts_test.go:437]: done

**Review by al1img — APPROVED** [review]

**Review by MykolaSolyanko — APPROVED** [review]

**Pull request merged**